0015-automatic-cluster-upgrades/0015-automatic-cluster-upgrades.md

```diff
@@ -0,0 +1,148 @@
+
+## Technical aspects
+
+We envision aforementioned concepts would become custom resource definitions and there will be additional controllers that would enhance and ensure the correct behavior.
+
+### Resources
+
+#### ClusterUpgrade
+
+Example resource:
+
+```yaml
+apiVersion: 
+kind: ClusterUpgrade
+metadata:
+  name: upgrade-to-release-v20
+  namespace: default
+spec:
+  clusterRef:
+    apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: foo01
+  release: v13.0.1
+  startTime: 2021-03-10T07:00:00Z
+status:
+  triggeredAt: 2021-03-10T10:00:00Z
+  state: finished
+  outcome: success
+  finishedAt: 2021-03-10T10:31:10Z
+  fromRelease: v13.0.0
+  toRelease: v13.0.1
+```
+
+#### MaintenanceSchedule
+
+Example resource:
+
+```yaml
+apiVersion: 
+kind: MaintenanceSchedule
+metadata:
+  name: working-hours-except-month-start
+  namespace: default
+spec:
+  maintenanceWindows:
+  - name: working-hours
+    startTime: 2021-03-10T07:00:00Z
+    endTime: 2021-03-10T17:00:00Z
+    recurrence: FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR
+  maintenanceExclusions:
+  - name: month-start
+    startTime: 2021-03-01T00:00:00Z
+    endTime: 2021-03-10T23:59:59Z
+    recurrence: FREQ=MONTHLY
+```
+
+This would allow upgrades to happen between 07:00 UTC and 17:00 UTC on Monday-Friday, except for days between 1st and 10th of a month.
+
+### Controllers
+
+#### Cluster upgrade executor
+
+There will be an operator that based on the `Cluster Upgrade` CRs will initiate the upgrades changing the labels on the specific CR(s), it could include Apps at some point though ideally would detach apps. totally from the cluster and. apps metadata and App Platform enforce the requirements. 
+
+This operator can also check on the `maintenance schedule` to verify the upgrade fulfil or not the criteria, and in case it does not, alert or set an appropriate status on the object status.
+
+Also the operator will handle the mechanics of the upgrade like which `MachinePool` is upgraded first or wait till one `MachinePool` is done successfully to continue with next one. There is already an [RFC](https://github.com/giantswarm/rfc/pull/17/files) up for this specific controller.
+
+#### Cluster upgrade scheduler
+
+Ideally to automate the whole process there will be an operator that creates all the `Cluster Upgrade` CRs when a new release of Giant Swarm is created (and/or possibly other events).
+
+#### Existing CAPI controllers
+
+There was [a proposal in CAPI upstream](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md) for the KubeadmControlPlane (later included in CAPZ components) to add `upgradeAfter` parameter to influence on the upgrades. The idea would be to create a new proposal that allows to point a new entity (`UpgradePolicy`) to extend the possibilities more than a single timestamp. Later controllers could leverage on that to allow changes or not to their reconciled resources.
```

Comment:
If I understand correctly, the `upgradeAfter` field was added to be able to trigger upgrades even when there is nothing to change (it's useful to re-create infrastructure and re-generate certs, for example). So I believe this wouldn't replace the `upgradeAfter` field, but it would be a new field. When the controller reconciling the CR's wants to upgrade something due to the `upgradeAfter` field, it would need to check the `UpgradePolicy`.

Reply:
yes it is the case